### ISSUE TITLE
Amazing post that aroused me a question.

### ISSUE THREAD

**Issue by johnunclesam** [equivalent/scrapbook2#10]
After I read this amazing post: https://github.com/equivalent/scrapbook2/blob/master/archive/blogs/2017-10-12-csrf-protection-on-single-page-app-api.md I'm wondering this: https://stackoverflow.com/questions/50159847/single-page-application-and-csrf-token.

Can you answer that?

**Comment by equivalent** [comment]
Hey @johnunclesam  thank you for you post. I'm glad you like my article. 

Sorry that I didn't answer the question earlier but I was hammered with work and unfortunately I'll not have time to fully answer it this weekend either (I'm on a trip) 

I'm bit rusty on CSRF topic and need to  remind myself on the topic before I can give valuable answer (as security answers are always complicated)

In short I would say just this: I'm using `protect_from_forgery with: :exception` for our SPA  API 

I'll fully explain myself once I'm back but here are some random resources that may help:


* https://nvisium.com/blog/2014/09/10/understanding-protectfromforgery.html
* https://blog.bigbinary.com/2016/04/06/rails-5-default-protect-from-forgery-prepend-false.html

> b.t.w. the new url of the article is https://blog.eq8.eu/article/csrf-protection-on-single-page-app-api.html 
> I changed the blog to Jekyll so I can write articles faster. Sorry for the problems

**Comment by equivalent** [comment]
Ah ok my bad now I understand what you are asking

it really depends on how the SPA is being rendered. 

## scenario 1 -  Rails  rendering the SPA 
 
e.g. angular-rails gem rendered via sprockets

So rails render application.html.erb with angular template and then the Angular JS just communicates with Rails via API

```
|-------------------|
| Rails             |
|    |-----------|  |
|    |SPA        |  |
|    |__________ |  |
__________________

```

## Scenario 2 -SPA that is rendered outside Rails

 (e.g. GitubPages Angualar JS App that is just communicating with pure Rails API, so Rails has nothing to do with redering the SPA)


```
 |--------|             |-------------|
 | SPA    |             | Rails          |
 |________|             |_________|

```


## Solutions

> It's 23:15 when I'm writing this response, I need to go to bed soon to wake up in few hours for a long road trip :smile:  ...I'll update my answer once I'm back, this is just quck B.S. top of my head

### For scenario 1

Funny storry is that I've looked into legacy coudebase of one project and we have exactly what the question is proposing https://stackoverflow.com/questions/50159847/single-page-application-and-csrf-token :

```
class ApplicationController < ActionController::Base

  after_action :set_csrf_cookie

  def set_csrf_cookie
    cookies["X-CSRF-Token"] = form_authenticity_token
  end

end
```

So I guess that's fine + `protect_from_forgery with: :null_session`

### For scenario 2 

I need to chat with our FE developer to remind myself what we are doing but I kindof know `protect_from_forgery with: :exception` plays important part in it ;) 

will reply back on Monday

**Comment by equivalent** [comment]
...now I'm not sure how will I fall asleep :)

**Comment by johnunclesam** [comment]
Everything ok in your trip?

**Comment by equivalent** [comment]
I've completely forget to answer this.  My apologies. I'll answer it now
on Stack overflow + here:

## recap on how CSRF works 

````
Given I'm a user of my-bank.com
And I'm signed in to my-bank.com (cookie session id / cookie JWT)
When I click on a malicious link (email, `<img src="...">`, 3rd party website) with url `http://my-bank?transfer_all_my_money_to_user=1234`
Then all my money ends up transfered to hacker user 1234
```

The whole point of CSRF token protection is that because of cookie
session_id/ cookie JWT is sent on every request we need one more
information (CSRF token.) that server will acknowledge as browser action
by user not by malicious link.

> Yes GET requests are not CSRF protected  in Rails because it's not
> common to do "actions" like "transfer all my money" with GET requests. 

So this leaves us to the point where should the CSRF be stored ? 

In HTML rendering Rails application  the CSRF token is rendered in
Rails form as a hidden field therefore it's send automatically when HTML forms are submitted.

```erb
<%= form_for @user do |f| >
  # ....
<% end %>
```

Rails HTML render will also render the CSRF token in the  `<meta name="csrf-token" ...>` tag 
so when it comes to AJAX  or Single page apps rendered from Rails we just
configure to pick the meta CSRF token and send it with the AJAX request.


```erb
<html>
  <head>
    <%= csrf_meta_tags %>
```

So thin CSRF token is generated once and used trough out of lifecycle of session:

> Rails will appear to generate a new CSRF token on every request, but it
> will accept any generated token from that session. In reality, it is
> just masking a single token using a one-time pad per request, in order
> to protect against SSL BREACH attack. More details at
> https://stackoverflow.com/a/49783739/2016618. You don't need to
> track/store these tokens.
>
> source: https://stackoverflow.com/a/50225227/473040 


When it comes to API only Ruby on Rails application we don't have this
luxory of `<mata>` tag .

### SPA scenario: Should I store CSRF in a cookie ?

Nature of cookies is that they are sent on every request BUT Rails is
not stupid and is checking the CSRF toker from `X-CSRF-TOKEN` header and
not cookie.

[source_code](https://github.com/rails/rails/blob/e7feaff70f13b56a0507e9f4dfaf3ebc361cb8e6/actionpack/lib/action_controller/metal/request_forgery_protection.rb#L102)

```ruby
# inside the `protect_from_forgery` method this method is activated:
# ...
def verified_request?
  !protect_against_forgery? || request.get? || request.head? ||
    valid_authenticity_token?(session, form_authenticity_param) ||
    valid_authenticity_token?(session, request.headers['X-CSRF-Token'])
end
# ...
```

 So in theory
if you store the CSRF token on  in a cookie
**that will not be read by server by default** it should be ok.

> It doesn't matter if the server (Rails) set the cookie or the SPA
> (Angular, React, ...)

The SPA needs to send the CSRF as a `header`.

Only way how the cookie CSRF store would be dangerous is if you write
your own CSRF token validation that would look something like this:


So don't do this:

```ruby
class MyController < ApplicationControlle
  def transfer_all_my_money
    raise 'CSRF token invalid' unless valid_authenticity_token?(session, cookie[:csrf_token])  # seriously never do this !!!!
    # ....
  end
end
```


> So rememmber, CSRF tokens should be sent via a header `X-CSRF-Token`. You need
> to configure your SPA to read the CSRF token from Local storage / Cookie
> and send it as this header.


So you can configure your Rails application to set CSRF token in a
cookie:

```ruby
class LoginController < ApplicationController
  def create
    if user_password_match
      # .... 
      cookie[:my_csrf_token]= form_authenticity_token
    end
  end
end
```

Or after every Post action

```ruby
class ApplicationController < ActionController::Base

  after_action :set_csrf_cookie

  def set_csrf_cookie
    cookies["X-CSRF-Token"] = form_authenticity_token
  end
end
```

I however prefer  CSRF as JSON API Body response

##  CSRF as JSON API Body response

Other way is to provide CSRF after login as a body response and  single page app
Angular, React, ...) stores it somewhere and send it with every request


####  SPA with session cookie cookie

If you use something like Devise for login, Devise will set session
cookie after login. What we may provide is a response for the login with
CSRF token:

```
curl POST  /login.json  -d "{"email":'equivalent@eq8.eu", "password":"Hello"}"  -H 'ContentType: application/json'

# Cookie with session_id was set

response:

{  "login": "ok", "csrf": 'yyyyyyyyy" }
```


#### JSON body CSRF - SPA with JWT token


```
curl POST  /login.json  -d "{"email":'equivalent@eq8.eu", "password":"Hello"}"  -H 'ContentType: application/json'

response:

{ "login": "ok", "jwt": "xxxxxxxxxxx.xxxxxxxx.xxxxx", "csrf": 'yyyyyyyyy" }
```


Now what is the idea ? I  think that the BE API should NOT be telling  SPA
how to store any information. This way we will give full responsibility
of "how to store" CSRF (and JWT) to SPA. Rails application will only look
at request headers `X-CSRF-TOKEN` for CSRF and `Authentication` for JWT
token.

```ruby
class ApplicationController
  protect_from_forgery # csrf header check
  before_action :authenticate

  # will check if `Authentication` header `Bearer: xxxxxxx.xxxxxx.xxxx` is valid
  def authenticate
    auth_header_value = request.header[:Authorization]
    jwt_token = auth_header_value.split(' ').last
    JWT.decode jwt_token, # ...
    # ...
  end
```

> also cookies are Domain locked if you need to create microservices
> that comunicate with multi domain with JWT you cannot do that with
> cookies.

**Comment by equivalent** [comment]
:thinking: actually let me turn this to an article and I'll post it to SO that way

**Comment by equivalent** [comment]
@johnunclesam  sorry for the 2 week delay. I forgot about this. 

I've formed my answer into this article :) https://blog.eq8.eu/article/rails-api-authentication-with-spa-csrf-tokens.html

Thx for the inspiration

**Comment by johnunclesam** [comment]
Fix the word `<mata>` and this `…)` in the article.

**Comment by johnunclesam** [comment]
Great article. But anyway I would like to know something.

Many times I see this:

```ruby
class ApplicationController < ActionController::Base

  after_action :set_csrf_cookie

  def set_csrf_cookie
    if current_user && protect_against_forgery?
      cookies["my_csrf_token"] = form_authenticity_token
    end
  end
end
```

It makes sense to have the code below?

```ruby
if current_user && protect_against_forgery?
```

**Comment by equivalent** [comment]
@johnunclesam  I've never used it like that so I cannot make statement that would be 100% secure.


In principle I cannot think about scenario where your code would be problem

But at the same time I don't see a problem to set the CSRF cookie on every request like this:

```
class ApplicationController < ActionController::Base

  after_action :set_csrf_cookie

  def set_csrf_cookie
      cookies["my_csrf_token"] = form_authenticity_token
  end
end
```

:thinking: 

honestly don't know. I definitely see what you are trying to archive  but both  solution (your and my code sample ) looks equally fine to me (but I may be wrong on this)


but just to be sure `protect_from_forgery` should be on globally.

```
class ApplicationController < ActionController::Base
   beforeprotect_from_forgery
end
```

and honestly there is no case I can think of where would developer need to do skip it with ` protect_from_forgery except: :create`  or `skip_before_action :verify_authenticity_token`. If that is the case why you use `protect_against_forgery?` in your code then you may have some security issues in your code.

**Comment by johnunclesam** [comment]
I think you're right.

It does not make sense to use `if current_user && protect_against_forgery?`.

**Comment by johnunclesam** [comment]
But maybe the `if current_user` can help prevent CSRF Breach attack? What do you think?

**Comment by equivalent** [comment]
well yes and no. CSRF Breach attack happens when attacker manage to figure out CSRF token from website compression (and yes even if the trafic is HTTPs) 

Quoting [this](https://www.adweek.com/digital/breach-csrf/) and [this](https://www.facebook.com/notes/protect-the-graph/preventing-a-breach-attack/1455331811373632) article

```
Under certain circumstances, attackers could figure out a victim’s secret token even when the communication was encrypted. Specifically, the attacker could discover information about the token because of the way that a Web page gets compressed.

Compression is a powerful way to speed up communication because repeated sections of text only need to get transmitted once. For example, if the word “Beetlejuice” is repeated three times in a Web page, the second and third instances get compressed into tiny shortcuts that refer to the first word. Similarly — and unfortunately — if any letters in the Web page match some of the letters in the CSRF token, compression makes the Web page smaller. This size difference is still present after the Web page is encrypted, and this effect could be used to reveal information about the CSRF token to any attackers who can see the size of the compressed Web page. A smart attacker could use a few hundred carefully crafted Web requests to figure out the entire token from the size alone.
```

That means that yes not exposing the CSRF token for every request (only for logged in users) is little bit helpful but still doesn't solve the issue (as attacker can still fig``ure out CSRF token of the current_user)


Therefore if your website is dealing with this level of concern around security you will have to rotate CSRF takens. Quoting same articles:

```
Facebook protects people against this threat with an extra layer of security inside the CSRF token. Before BREACH was invented, the token was rotated once daily. The CSRF token contained a truncated SHA-2 hash that incorporated the account ID and current date. A person with three Facebook sessions within a single day would have received an identical CSRF token each time, (e.g., AQAOQ2sf, AQAOQ2sf, and AQAOQ2sf). Now our system replaces the token with a new one every time it is requested. Three different sessions use three different CSRF tokens, (e.g., AQGSRmYTeFnr, AQFkqN92V78v, and AQHouyYa35iv). These new tokens are generated by introducing a random 24-bit salt. The salt is the last four letters at the end of the token and is also included within the hash, which eliminates all repetition anywhere in the token. After a new token is issued, the previous tokens still remain valid for a couple of days, resulting in multiple tokens being permissible simultaneously
```

Therefore you are doing that with:

```
class ApplicationController < ActionController::Base

  after_action :set_csrf_cookie  # important

  def set_csrf_cookie
    if current_user && protect_against_forgery?   # not important
      cookies["my_csrf_token"] = form_authenticity_token   # important
    end
  end
end
```

So you are actually implementing 2 security steps here:

* not exposing CSRF tokens to public users (` if current_user `) => decreasing the knowledge for hacker on how your CSRF tokens look like. This make sense if you are building "internal" application where note anyone can create an account :+1:. But if I can come to your website and create an account then this is not necessarily effective :-1:  
* rotating CSRF token with  ` after_action :set_csrf_cookie ` and `cookies["my_csrf_token"] = form_authenticity_token` :+1:   But the question is it as secured rotation as Facebook is providing :thinking: ?

So in conclusion

If you really need to bulletproof the security I would recommend investigating Rails source code around what  CSRF  token https://github.com/rails/rails/blob/e7feaff70f13b56a0507e9f4dfaf3ebc361cb8e6/actionpack/lib/action_controller/metal/request_forgery_protection.rb#L263
 
It's not as sophisticated as Facebook is proposing but still provides some good tricks to prevent BREACH yet if I was a Bank and had to have bulletproof CSRF I would investigate a way how to add one more CSRF like token that is generated per user base (so Rails CSRF + my custom CSRF just to be sure so I'm super paranoid) 

If this CSRF protection is for my personal Blog website I wouldn't bother with that and just use Rails native CSRF.

**Comment by equivalent** [comment]
:thinking: actually @johnunclesam Thank you again, now I realize that I've forgot to mention that in the article. Updating now.

**Comment by alitinkerlust** [comment]
Hey @equivalent! 

Thanks for your wonderful post.

In Scenario 2 (SPA that is rendered outside Rails):

- User visits a Login Page on the SPA the first time.
- User does not have a CSRF token in the cookie
- User would like to login `POST /login`

If we `protect_from_forgery` in ApplicationController, it will not be able to verify the CSRF token, since it does not exist.

Does this mean we will have to do a `GET /csrf` on the SPA if there it does not have a CSRF token in the cookie?

Or do we skip `protect_from_forgery` for `POST /login` controller actions?

**Comment by alialjunied** [comment]
Hey @equivalent!

Thanks for your wonderful post.

In Scenario 2 (SPA that is rendered outside Rails):

- User visits a Login Page on the SPA the first time.
- User does not have a CSRF token in the cookie
- User would like to login `POST /login`

If we `protect_from_forgery` in `ApplicationController`, it will not be able to verify the CSRF token, since it does not exist.

Does this mean we will have to do a `GET /csrf `on the SPA before `POST /csrf`?

Or do we skip `protect_from_forgery` for `POST /login` controller actions?
